lib/benchmark/benchmark.go

```diff
@@ -0,0 +1,297 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package benchmark package provides tools to run progressive or independent benchmarks against teleport services.
+package benchmark
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/HdrHistogram/hdrhistogram-go"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// minValue is the min millisecond recorded for histogram
+	minValue = 1
+	// maxValue is the max millisecond recorded for histogram
+	maxValue = 60000
+	// significantFigures is the precision of the values
+	significantFigures = 3
+	// pauseTimeBetweenBenchmarks is the time to pause between each benchmark
+	pauseTimeBetweenBenchmarks = time.Second * 5
+)
+
+// Config specifies benchmark requests to run
+type Config struct {
+	// Rate is requests per second origination rate
+	Rate int
+	// Command is a command to run
+	Command []string
+	// Interactive turns on interactive sessions
+	Interactive bool
+	// MinimumWindow is the min duration
+	MinimumWindow time.Duration
+	// MinimumMeasurments is the min amount of requests
+	MinimumMeasurements int
+}
+
+// Result is a result of the benchmark
+type Result struct {
+	// RequestsOriginated is amount of requests originated
+	RequestsOriginated int
+	// RequestsFailed is amount of requests failed
+	RequestsFailed int
+	// Histogram holds the response duration values
+	Histogram *hdrhistogram.Histogram
+	// LastError contains last recorded error
+	LastError error
+	// Duration it takes for the whole benchmark to run
+	Duration time.Duration
+}
+
+// Run is used to run the benchmarks, it is given a generator, command to run,
+// a host, host login, and proxy. If host login or proxy is an empty string, it will
+// use the default login
+func Run(ctx context.Context, lg *Linear, cmd, host, login, proxy string) ([]Result, error) {
+	c := strings.Split(cmd, " ")
+	lg.config = &Config{Command: c}
+	if err := validateConfig(lg); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tc, err := makeTeleportClient(host, login, proxy)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		exitSignals := make(chan os.Signal, 1)
+		signal.Notify(exitSignals, syscall.SIGTERM, syscall.SIGINT)
+		defer signal.Stop(exitSignals)
+		sig := <-exitSignals
+		logrus.Debugf("signal: %v", sig)
+		cancel()
+	}()
+	var results []Result
+	sleep := false
+	for {
```

Comment:
Will this work:
```go
	benchmarkConfig := lg.GetBenchmark()
	for benchmarkConfig != nil {
		result, err := benchmarkC.Benchmark(ctx, tc)
		if err != nil {
			return results, trace.Wrap(err)
		}
		results = append(results, result)
		fmt.Printf("current generation requests: %v, duration: %v\n", result.RequestsOriginated, result.Duration)
		select {
			case <-time.After(pauseTimeBetweenBenchmarks):
			case <-ctx.Done():
				return results, trace.ConnectionProblem(ctx.Err(), "context canceled or timed out")
		}
		benchmarkConfig := lg.GetBenchmark()
	}
```

Reply:
yes, this totally works actually

Reply:
Just noted a typo:
```
benchmarkConfig = lg.GetBenchmark()
```
inside the loop mustn't shadow.